bark_ml/library_wrappers/lib_fqf_iqn_qrdqn/agent/base_agent.py

```diff
@@ -0,0 +1,267 @@
+        else:
+            self.memory = LazyMultiStepMemory(
+                self._params["ML"]["BaseAgent"]["MemorySize",
+                                                "", 10**6], self.env.observation_space.shape,
+                self.device, self._params["ML"]["BaseAgent"]["Gamma", "", 0.99], self._params["ML"]["BaseAgent"]["Multi_step", "", 1])
+
+    def run(self):
+        while True:
+            self.train_episode()
+            if self.steps > self.num_steps:
+                break
+
+    def is_update(self):
+        return self.steps % self.update_interval == 0\
+            and self.steps >= self.start_steps
+
+    def is_random(self, eval=False):
+        # Use e-greedy for evaluation.
+        if self.steps < self.start_steps:
+            return True
+        if eval:
+            return np.random.rand() < self.epsilon_eval
+        if self.noisy_net:
+            return False
+        return np.random.rand() < self.epsilon_train.get()
+
+    def update_target(self):
+        self.target_net.load_state_dict(
+            self.online_net.state_dict())
+
+    def explore(self):
+        # Act with randomness.
+        action = self.env.action_space.sample()
+        return action
+
+    def exploit(self, state):
+        # Act without randomness.
+        #state = torch.Tensor(state).unsqueeze(0).to(self.device).float() 
+        actions = self.calculate_actions(state).argmax().item()
+        return actions
+
+    def calculate_actions(self, state):
+        # Act without randomness.
+        state = torch.Tensor(state).unsqueeze(0).to(self.device).float() 
+        with torch.no_grad():
+            actions = self.online_net(states=state)
+        return actions
+
+    @abstractmethod
+    def learn(self):
+        pass
+
+    def save_models(self, save_dir):
+        if not os.path.exists(save_dir):
+            os.makedirs(save_dir)
+        torch.save(
+            self.online_net.state_dict(),
+            os.path.join(save_dir, 'online_net.pth'))
+        torch.save(
+            self.target_net.state_dict(),
+            os.path.join(save_dir, 'target_net.pth'))
+        online_net_script = torch.jit.script(self.online_net)
+        online_net_script.save(os.path.join(save_dir, 'online_net_script.pt'))
+
+    def load_models(self, save_dir):
+        self.online_net.load_state_dict(torch.load(
+            os.path.join(save_dir, 'online_net.pth')))
+        self.target_net.load_state_dict(torch.load(
+            os.path.join(save_dir, 'target_net.pth')))
+
+    def visualize(self, num_episodes=5):
+        for _ in range(0, num_episodes):
+            state = self.env.reset()
+            done = False
+            while (not done):
+                action = self.exploit(state)
+                next_state, reward, done, _ = self.env.step(action)
+                self.env.render()
+                state = next_state
+
+    def train_episode(self):
+        self.online_net.train()
```

Comment:
do you not want to split this in a runner and agent class, see tf-agents

Reply:
The architecture is not modified to maintain compatibility with the [repository](https://github.com/ku2482/fqf-iqn-qrdqn.pytorch) on which it's based upon to simplify integrating new upstream changes.